### PR TITLE
Fix Flux.concat batchSize over-producing

### LINKED ISSUE #2398: Stream from Flux::toStream() makes unnecessary upstream subscriptions.
The following code mimics the logic we have to fetch pages from a remote service. The service returns result over multiple pages. Each page from the service will have the ID of the next Page.

We use the `concat` Operator to follow the pages and, serially fetch them.

```java
 public class PageSource {
    private final int maxPages;
    private final AtomicReference<Page> currentPageRef;

    public PageSource(int maxPages) {
        this.maxPages = maxPages;
        this.currentPageRef = new AtomicReference<>();
    }

    public Flux<Page> getPageFlux() {
        return this.concatPages();
    }

    private Flux<Page> concatPages() {
        Page currentPage = this.currentPageRef.get();
        if (currentPage != null && currentPage.pageId == this.maxPages) {
            return Flux.empty();
        } else {
            return this.getNextPage(currentPage)
                    .doOnNext(nextPage -> {
                        this.currentPageRef.set(nextPage);
                    })
                    .concatWith(Flux.defer(() -> this.concatPages()));
        }
    }

     private Mono<Page> getNextPage(Page currentPage) {
        if (currentPage == null) {
            return Mono.defer(() -> {
                System.out.println("Mocked network call for Page_0");
                return Mono.just(new Page(0));
            });
        } else {
            System.out.println("Mocked network call for Page_" + currentPage.nextPageId);
            return Mono.just(new Page(currentPage.nextPageId));
        }
     }

     static class Page {
         public final int pageId;
         public final List<String> items;
         public final int nextPageId;

         public Page(int pageId) {
             this.pageId = pageId;
             this.items = new ArrayList<>();
             this.items.add(String.format("Page_%s_item_1", pageId));
             this.items.add(String.format("Page_%s_item_2", pageId));
             this.items.add(String.format("Page_%s_item_3", pageId));
             this.items.add(String.format("Page_%s_item_4", pageId));
             this.items.add(String.format("Page_%s_item_5", pageId));
             this.nextPageId = this.pageId + 1;
         }
     }
 }
```

The Concat operator ensures the expensive network calls are happening only when needed. In the below test, we can see that as expected, there is only one page fetched.

test_1:

```
 final int TOTAL_PAGES = 500;
 PageSource dataSource = new PageSource(TOTAL_PAGES);
 dataSource.getPageFlux().blockFirst();
```

output_1:

```
Mocked network call for Page_0
```

But we have an issue with the `toStream` operator and `findFirst()` terminal operator.

here we can see that though we asked for 1 item it made 3 network calls. (Note that we're also using batch_size as 1).

```java
 final int TOTAL_PAGES = 500;
 final int BATCH_SIZE = 1;
 PageSource dataSource = new PageSource(TOTAL_PAGES);
 dataSource
         .getPageFlux()
         .toStream(BATCH_SIZE)
         .findFirst();
```

```
Mocked network call for Page_0
Mocked network call for Page_1
Mocked network call for Page_2
```

We don't really need the above page code to repro this, I just used it to demonstrate the real use case, below code can also repro this.

```java
 Flux<PageSource.Page> pageFlux = Mono.defer(() -> {
     System.out.println("Getting page_0");
     return Mono.just(new PageSource.Page(0));
 }).concatWith(Mono.defer(() -> {
     System.out.println("Getting page_1");
     return Mono.just(new PageSource.Page(1));
 })).concatWith(Mono.defer(() -> {
     System.out.println("Getting page_2");
     return Mono.just(new PageSource.Page(2));
 })).concatWith(Mono.defer(() -> {
     System.out.println("Getting page_3");
     return Mono.just(new PageSource.Page(3));
 })).concatWith(Mono.defer(() -> {
    System.out.println("Getting page_4");
    return Mono.just(new PageSource.Page(3));
 })).concatWith(Mono.defer(() -> {
     System.out.println("Getting page_5");
     return Mono.just(new PageSource.Page(3));
 }));

pageFlux.toStream(1).findFirst();
```

```
Getting page_0
Getting page_1
Getting page_2
```

This is impacting our customers. The customer uses this in multiple places and the extra 3 calls increase total network calls: 3x. Any help to fix this will be really appreciated.

Thanks!

### PR DESCRIPTION
**Pull request by camsteffen**
My first contribution! 😃 Check my math since I'm not super familiar with the internals.

Fixes #2398

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@camsteffen Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/reactor?repositoryId=reactor/reactor-core&pullRequestId=2402)!

[Click here](https://cla.pivotal.io/sync/reactor?repositoryId=reactor/reactor-core&pullRequestId=2402) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@camsteffen Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/reactor?repositoryId=reactor/reactor-core&pullRequestId=2402)!

File: reactor-core/src/test/java/reactor/core/publisher/FluxConcatArrayTest.java [hunk 1]
```diff
@@ -313,4 +315,13 @@ public void scanDelayErrorSubscriber() {
 		assertThat(test.scan(Scannable.Attr.CANCELLED)).isTrue();
 	}
 
+	@Test
+	public void iterableBatchSize() {
+		Flux<Integer> flux = Flux.concat(
+				Mono.just(0),
+				Mono.just(1).doOnSubscribe(s -> fail("batchSize did not prevent subscription")));
```
Comment (camsteffen) [inline reactor-core/src/test/java/reactor/core/publisher/FluxConcatArrayTest.java:322]: Is there a better way to do this?

**Comment by camsteffen** [comment]
Welp looks like the fix is wrong ☹️

File: reactor-core/src/test/java/reactor/core/publisher/FluxConcatArrayTest.java (hunk 1 above)
Comment (simonbasle) [reply reactor-core/src/test/java/reactor/core/publisher/FluxConcatArrayTest.java:322]: you could look into replacing the `Mono.just(1)` with a `TestPublisher`, instantiated in a dedicated variable. The `TestPublisher` can then be asserted for the absence of subscription.

**Comment by camsteffen** [comment]
I'm afraid I've bitten of more than I can chew here. I've learned some things about the internals in trying to fix this, but it's rather overwhelming. I found out that concatMap with prefetch=0 also will have eager prefetching after `toIterable(1)`. Perhaps there are other cases as well. It seems like eager prefetching is just deeply baked in.

**Pull request closed without merging**